reconcile.go

```diff
@@ -234,54 +244,102 @@ func (a *Autopilot) pruneDeadServers() error {
 
 	state := a.GetState()
 
-	failed, voters, err := a.getFailedServers()
+	failed, vr, err := a.getFailedServers()
 	if err != nil || failed == nil {
 		return err
 	}
 
 	failed = a.promoter.FilterFailedServerRemovals(conf, state, failed)
 
-	// remove failed non voting servers
-	for _, srv := range failed.FailedNonVoters {
-		a.logger.Info("Attempting removal of failed server node", "id", srv.ID, "name", srv.Name, "address", srv.Address)
-		a.delegate.RemoveFailedServer(srv)
+	// Remove servers in order of increasing precedence
+	// Update the voter registry after each batch is processed
+
+	// remove stale non-voters
+	toRemove := a.adjudicateRemoval(failed.StaleNonVoters, vr)
+	if err = a.removeStaleServers(toRemove); err != nil {
+		return err
 	}
+	vr.RemoveAll(toRemove)
 
-	// remove stale non voters
-	for _, id := range failed.StaleNonVoters {
-		a.logger.Debug("removing stale raft server from configuration", "id", id)
-		if err := a.removeServer(id); err != nil {
-			return err
-		}
+	// Remove stale voters
+	toRemove = a.adjudicateRemoval(failed.StaleVoters, vr)
+	if err = a.removeStaleServers(toRemove); err != nil {
+		return err
 	}
+	vr.RemoveAll(toRemove)
 
-	maxRemoval := (voters - 1) / 2
+	// remove failed non-voters
+	failedNonVoters := vr.Filter(failed.FailedNonVoters)
```

Comment:
What is the purpose of this filter call?

Reply:
Just to filter the `[]*Server` objects to only return the `raft.ServerID`s of the ones we want.

Reply:
Which ones do we want?  I'm confused because it's just ensuring that these nodes are present in the map, and I'm not sure why that's needed.

Reply:
We want the collection of server ID (`[]raft.ServerID`), on `FailedServers` object, the 'stale' fields, `StaleNonVoters` and `StaleVoters` are both just collections (slices) or the ID. But `FailedNonVoters` and `FailedVoters` are collections of server objects. So that filter is just a way to get the IDs, but ensuring the voter registry knows about them (as we pull voter eligibility info from the registry inside the adjudicate func).